src/main/scala/mesosphere/marathon/core/task/Task.scala

```diff
@@ -67,6 +64,211 @@ case class Task(
+
+    sealed trait State {
+      /** Defines when this state should time out and for which reason */
+      def timeout: Option[Timeout]
+    }
+    object State {
+      /** A newly reserved Task */
+      case class New(timeout: Option[Timeout]) extends State
+      /** A ResidentTask that has been running before but terminated and can be relaunched */
+      case class Suspended(timeout: Option[Timeout]) extends State
+      /** A ResidentTask whose reservation and persistent volumes are being destroyed */
+      case class Garbage(timeout: Option[Timeout]) extends State
+      /** An unknown task created because of unknown reservations/persistent volumes */
+      case class Unknown(timeout: Option[Timeout]) extends State
+    }
+
+    /**
+      * A timeout that eventually leads to a state transition
+      * @param initiated When this timeout was setup
+      * @param deadline When this timeout should become effective
+      * @param reason The reason why this timeout was set up
+      */
+    case class Timeout(initiated: Timestamp, deadline: Timestamp, reason: Timeout.Reason)
+
+    object Timeout {
+      sealed trait Reason
+      object Reason {
+        /** A timeout because the task could not be relaunched */
+        case object RelaunchEscalationTimeout extends Reason
+        /** A timeout because we got no ack for reserved resources or persistent volumes */
+        case object ReservationTimeout extends Reason
+      }
+    }
+  }
+
+  case class LaunchedOnReservation(
+      taskId: Task.Id,
+      agentInfo: AgentInfo,
+      appVersion: Timestamp,
+      status: Status,
+      networking: Networking,
+      reservation: ReservationWithVolumes) extends Task {
+
+    private[this] val log = LoggerFactory.getLogger(getClass)
+
+    override lazy val reservationWithVolumes: Option[ReservationWithVolumes] = Some(reservation)
+
+    override lazy val launched: Option[Launched] = Some(Task.Launched(appVersion, status, networking))
+
+    private[this] def hasStartedRunning: Boolean = status.startedAt.isDefined
+
+    override def update(update: TaskStateOp): TaskStateChange = update match {
+      // case 1: now running
+      case TaskStateOp.MesosUpdate(MarathonTaskStatus.Running(mesosStatus), now) if !hasStartedRunning =>
+        val updated = copy(
+          status = status.copy(
+            startedAt = Some(now),
+            mesosStatus = mesosStatus))
+        TaskStateChange.Update(updated)
+
+      // case 2: terminal
+      // FIXME (3221): handle task_lost, kill etc differently and set appropriate timeouts (if any)
+      case TaskStateOp.MesosUpdate(MarathonTaskStatus.Terminal(_), now) =>
+        TaskStateChange.Update(Task.Reserved(
+          taskId = taskId,
+          agentInfo = agentInfo,
+          state = Task.Reserved.State.Suspended(timeout = None),
+          reservation = reservation
+        ))
+
+      // case 3: health updated
+      case TaskStateOp.MesosUpdate(taskStatus, now) =>
```

Comment:
We could also get other states here when reconciling. The code than ignores the updates, which should be mostly fine but I am not certain. Not sure if we should change something here...